Fix flaky failure of TestWebKitAPI.ServiceWorkers.CacheStorageNetworkProcessCrash

RupinMittal opened the pull request:
#### ede345f72045232e2a234574f3824cabd6bd1016
<pre>
Fix flaky failure of TestWebKitAPI.ServiceWorkers.CacheStorageNetworkProcessCrash
<a href="https://bugs.webkit.org/show_bug.cgi?id=280208">https://bugs.webkit.org/show_bug.cgi?id=280208</a>
<a href="https://rdar.apple.com/136529803">rdar://136529803</a>

Reviewed by Youenn Fablet and Sihui Liu.

The API test TestWebKitAPI.ServiceWorkers.CacheStorageNetworkProcessCrash attempts to:
1. Create a cache
2. Terminate the network process
3. Check that trying to use the cache results in an error (indicating that
   the cache was invalidated since the network process it was attached to
   has been terminated).

The problem is that the call to use the cache in step 3 (DOMCache::match)
goes through the web process (WebCacheStorageConnection). But the call
to terminate the network process doesn&apos;t immediately notify the web process
that the network process has been terminated (its async). So if the call to
use the cache happens before the web process is aware, the call will try to
use the old network process, eventually realize that this is invalid, and
fail with an internal error.

So there is an inherent race condition.

The old test tried to get around this by first trying to open caches,
hoping that the web process would be notified of the network process&apos;
termination at some point within these calls and then match could be tested
as expected. But it didn&apos;t handle the internal error that flakily occurs
as described above.

So we add the proper error handling--if a call to open() fails, we just
retry instead of causing the test to fail.

Additionally, we have a limit of 10 calls to open() so that the test
does not keep trying until timeout. If none of these 10 calls succeed
(which is possible given the race-condition), the test will be flaky again
and we&apos;ll need to increase the number of calls. (10 seems appropriate
for now given that the test usually takes only 1 or 2 to pass locally.)

Once a call to open() succeeds, we know that the web process has realized
the network process was terminated and has opened a new connection to a new
network process. Then, we make the call to use the original cache and check
if it is invalid.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
((ServiceWorkers, CacheStorageNetworkProcessCrash)):
((ServiceWorkers, DISABLED_CacheStorageNetworkProcessCrash)(ServiceWorkers, CacheStorageNetworkProcessCrash)): Deleted.

Canonical link: <a href="https://commits.webkit.org/286678@main">https://commits.webkit.org/286678@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7fc249b1280dd2b56e6820119d49103a155bba63

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76707 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55742 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29613 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81239 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27983 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64884 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4035 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60116 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18212 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79774 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50059 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65859 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40438 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47460 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23353 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26307 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23683 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82684 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4083 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2696 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68398 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4236 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65831 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67649 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/16877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11630 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9708 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11873 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4030 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4053 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7483 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5811 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -3415,48 +3415,55 @@ HTTPServer server([&] (Connection connection) {
+{
+    return new Promise((resolve, reject) => {
+        self.caches.open("cache" + nextCacheNumber++).then(() => {
+            cache1.match('my request').then(() => {
```
`match` should not be in the loop based on what the comment below said.
It should probably be something like:
```
// Open
const maxAttempt = 10;
let count  = 0;
let newCache = null;
while (count < maxAttempt) {
    try {
       newCache = await self.caches.open(cacheName);
    } catch(e) {
       ++count;
    }
}

if (!newCache) {
    sendFailureMessage();
    return;
}

// Match
cache1.match()...
```

RupinMittal replied:
Good suggestion--the code does look nicer that way. I've updated and uploaded it again.

But why do we want an attempt limit? This would make the test non-deterministic, which is exactly the problem we're trying to solve. Isn't it better to keep looping until the open call succeeds? That way, the test won't be flaky.

szewai replied:
Based on what you showed me, I think the limit is not what makes the test non-deterministic, but original test does not expect all `open()` calls to succeed? 
We usually have a max attempt in test because we don't want test to keep trying until timeout. First, normally it does not need 30s to prove something is wrong. Second it would be hard to debug timeout (we won't know if it's because some promise does not resolve or because of the loop). Third timeout would slow down the whole test run.
If you have concern, we could start with some big number like 50 (but 10 should be enough I think, based on the original test). Also, we want a different error message (than match fail) when the test fails because of 50 opens fail, as in the code above.

RupinMittal replied:
It is true that the original test was failing because it doesn't expect all open() calls to succeed. But it's also possible that all of the 10 calls fail since there's no guarantee on when the web process is notified--so the test would still be non-deterministic. (These are the two issues I was trying to fix). 

So I thought this loop would be good since it runs exactly the number of open() calls to make the test succeed and locally it has been running quickly (<1 second).

But I did put the limit in since you mentioned that we don't want the test to keep trying until timeout in case something is wrong. And I added a comment so that if the test shows up as flaky again, we know to increase the number of attempts :)

Let me know if it looks good now--thank you!

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -3415,48 +3415,46 @@ HTTPServer server([&] (Connection connection) {
-    }, e => {
-        webkit.messageHandlers.sw.postMessage("check failed with " + e);
-    });
+    var nextCacheNumber = 2;
```
Even though the current implementation does not care much about the number of caches, can we increase it to 10 like done previously?

RupinMittal replied:
Hello! I'm not quite sure I understand. Do you mean we should create a minimum of 10 caches? Or do you mean set a limit on the number to be 10 as Sihui described? If so, I've done that in the most recent iteration :)

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -3415,48 +3415,57 @@ HTTPServer server([&] (Connection connection) {
-        webkit.messageHandlers.sw.postMessage("check failed with " + e);
-    });
+    const maxAttempts = 10;
+    let attempt = 1;
```
Nit: there is no need for additional parameter, you could either compare `nextCacheNumber` with const maxAttempts (i.e. `nextCacheNumber <= maxAttempts`) or just do `while (--remainingAttempts) {  try { } catch(e) { }}`

RupinMittal replied:
Good point. Since we only create one cache, we don't need each attempt to try creating one with a unique name. We can use the same name on each try. So I just removed `nextCacheNumber`.

RupinMittal committed:
Fix flaky failure of TestWebKitAPI.ServiceWorkers.CacheStorageNetworkProcessCrash
https://bugs.webkit.org/show_bug.cgi?id=280208
rdar://136529803

Reviewed by Youenn Fablet and Sihui Liu.

The API test TestWebKitAPI.ServiceWorkers.CacheStorageNetworkProcessCrash attempts to:
1. Create a cache
2. Terminate the network process
3. Check that trying to use the cache results in an error (indicating that
   the cache was invalidated since the network process it was attached to
   has been terminated).

The problem is that the call to use the cache in step 3 (DOMCache::match)
goes through the web process (WebCacheStorageConnection). But the call
to terminate the network process doesn't immediately notify the web process
that the network process has been terminated (its async). So if the call to
use the cache happens before the web process is aware, the call will try to
use the old network process, eventually realize that this is invalid, and
fail with an internal error.

So there is an inherent race condition.

The old test tried to get around this by first trying to open caches,
hoping that the web process would be notified of the network process'
termination at some point within these calls and then match could be tested
as expected. But it didn't handle the internal error that flakily occurs
as described above.

So we add the proper error handling--if a call to open() fails, we just
retry instead of causing the test to fail.

Additionally, we have a limit of 10 calls to open() so that the test
does not keep trying until timeout. If none of these 10 calls succeed
(which is possible given the race-condition), the test will be flaky again
and we'll need to increase the number of calls. (10 seems appropriate
for now given that the test usually takes only 1 or 2 to pass locally.)

Once a call to open() succeeds, we know that the web process has realized
the network process was terminated and has opened a new connection to a new
network process. Then, we make the call to use the original cache and check
if it is invalid.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
((ServiceWorkers, CacheStorageNetworkProcessCrash)):
((ServiceWorkers, DISABLED_CacheStorageNetworkProcessCrash)(ServiceWorkers, CacheStorageNetworkProcessCrash)): Deleted.

Canonical link: https://commits.webkit.org/286678@main
